src/connector/src/sink/remote.rs

```diff
@@ -0,0 +1,574 @@
+}
+
+#[derive(Clone, Debug)]
+pub struct RemoteConfig {
+    pub sink_type: String,
+    pub properties: HashMap<String, String>,
+}
+
+impl RemoteConfig {
+    pub fn from_hashmap(values: HashMap<String, String>) -> Result<Self> {
+        let sink_type = values
+            .get("connector")
+            .expect("sink type must be specified")
+            .to_string();
+
+        if !is_valid_remote_sink(sink_type.clone()) {
+            return Err(SinkError::Config(format!("invalid sink type: {sink_type}")));
+        }
+
+        Ok(RemoteConfig {
+            sink_type,
+            properties: values,
+        })
+    }
+}
+
+#[derive(Debug)]
+pub enum RemoteSinkResponseImpl {
+    Grpc(Streaming<SinkResponse>),
+    Receiver(UnboundedReceiver<SinkResponse>),
+}
+
+impl RemoteSinkResponseImpl {
+    pub async fn on_response_ok(&mut self) -> std::result::Result<SinkResponse, SinkError> {
+        return match self {
+            RemoteSinkResponseImpl::Grpc(ref mut response) => response
+                .next()
+                .await
+                .unwrap_or_else(|| Err(Status::cancelled("response stream closed unexpectedly")))
+                .map_err(|e| SinkError::Remote(e.message().to_string())),
+            RemoteSinkResponseImpl::Receiver(ref mut receiver) => receiver
+                .recv()
+                .await
+                .ok_or_else(|| SinkError::Remote("response stream closed unexpectedly".to_string()))
+        };
+    }
+}
+
+#[derive(Debug)]
+pub enum ConnectorClientImpl {
+    Grpc(ConnectorServiceClient<Channel>),
+    Mock,
+}
+
+#[derive(Debug)]
+pub struct RemoteSink {
+    pub sink_type: String,
+    properties: HashMap<String, String>,
+    epoch: Option<u64>,
+    batch_id: u64,
+    schema: Schema,
+    client: ConnectorClientImpl,
+    pub request_sender: Option<UnboundedSender<SinkTask>>,
+    pub response_stream: RemoteSinkResponseImpl,
+}
+
+pub struct RemoteSinkParams {
+    pub connector_addr: String,
+}
+
+impl RemoteSink {
+    pub async fn new(
+        config: RemoteConfig,
+        schema: Schema,
+        pk_indices: Vec<usize>,
+        sink_params: RemoteSinkParams,
+    ) -> Result<Self> {
+        let channel = Endpoint::from_shared(format!("http://{}", sink_params.connector_addr))
+            .map_err(|e| SinkError::Remote(format!("failed to connect channel: {:?}", e)))?
+            .initial_connection_window_size(MAX_CONNECTION_WINDOW_SIZE)
+            .initial_stream_window_size(STREAM_WINDOW_SIZE)
+            .tcp_nodelay(true)
+            .connect_timeout(Duration::from_secs(5))
+            .connect()
+            .await
+            .map_err(|e| SinkError::Remote(format!("failed to connect channel: {:?}", e)))?; // create client and start sink
+        let mut client = ConnectorServiceClient::new(channel);
+
+        let (request_sender, request_receiver) = mpsc::unbounded_channel::<SinkTask>();
+        request_sender
```

Comment:
May add a comment here saying that it is important to send the request first before initiating the stream since the response of creating stream will blocking-wait on the first request.